File: 49.py
from collections import defaultdict


class Solution:
    def groupAnagrams(self,strs: list[str]) -> list[list[str]]:
        z = []
        d = defaultdict(list)
        ans = []
        for i in strs:
            if sorted(i) not in z:
                z.append(sorted(i))
        for i,j in enumerate(strs):
            # print(sorted(j))
            if sorted(j) in z:
                d[str(sorted(j))].append(j)
        for i in z:
            ans.append(d[str(i)])
        return sorted(ans, key=len)

File: 49.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  let z = [];
  let d = {};
  let ans = [];
  for (let i of strs) {
    const temp = i.split("").sort().join("");
    if (!d[temp]) {
      d[temp] = [];
      z.push(temp);
    }
    d[temp].push(i);
  }
  for (let j of z) {
    ans.push(d[j]);
  }
  return ans.sort((a, b) => a.length - b.length);
};

const input = ["eat", "tea", "tan", "ate", "nat", "bat"];
console.log(groupAnagrams(input));
